docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+
+1) We want to run evaluations in an online fashion so we can use those evaluations for triggering policies. For example, we may want to trigger a policy that will retrain a model if the model's performance degrades below a certain threshold.
+
+2) We want to run evaluations in an offline fashion for experiments so we determine the performance of a model **[1] at a certain point in time** as well as the **[2] total pipeline performance** aggregated over the whole time of the pipeline run (= pipeline performance).
+
+## Setting
+
+We are looking at a completed pipeline run that produced multiple triggers and therefore multiple models. We want to evaluate the performance of each model.
+
+```mermaid
+gantt
+    title A Gantt Diagram
+    dateFormat YYYY-MM-DD
+    axisFormat %Y
+    section models
+        Trigger1 :milestone, 1930, 0d
+        model 1          :a1, 1930, 10y
+        Trigger2 :milestone, :after a1
+        model 2          :a2, after a1, 5y
+        Trigger3 :milestone, :after a2
+        model 3          :a3, after a2, 5y
+        Trigger4 :milestone, :after a3
+        model 4          :a4, after a3, 25y
+        Trigger5 :milestone, :after a4
+        model 5          :a5, after a4, 25y
+        Trigger6 :milestone, :after a5
+        model 6          :a6, after a5, 5y
+        Trigger7 :milestone, :after a6
+        model 7          :a7, after a6, 5y
+```
+
+## Evaluation methodology
+
+An evaluation methodology consists of three main components:
+
+1) The evaluation metric (e.g. F1, accuracy, precision, recall, etc.)
+2) The evaluation data (set of samples out of the evaluation split used for single model evaluation)
+3) The points in time when the evaluations are run
+
+### Use case 1: online evaluation
+
+In this case we have to be extremely cautious to keep evaluation costs low as the are part of the trigger policy evaluation costs (or at least part of the core pipeline that cannot be disregarded like the experiment model evaluations).
+
+During the core pipeline execution (not including experiment mode pipeline steps), we cannot access future data for evaluation. Also we don't have an evaluation split available. Therefore, we need to use the training data up to the current point in time for evaluation. This is not ideal but it is the only way to evaluate the model in an online fashion.
+
+> The question of how far into the past we want to includes samples remains. More thought is needed before continuing on this.
+
+### Use case 2: offline evaluation
+
+In experiment mode we can potentially utilize the full evaluation split over the whole dataset time range. Being able to "look into the future" is important to determine the performance of a model.
+
+#### [2] Total pipeline performance
+
+We argue that the performance of pipeline w.r.t. a certain evaluation metric is the weighted average of the performance of the most up to date model at each point in time over the whole time range of the pipeline run.
+
```

Comment:
I am not convinced that we need to do a weighted average. It's also just fine (straightforward and intuitive) to plot a curve with each evaluation in time.

In the end we want to publish a paper. I think the paper reader would prefer a straightforward, easy-to-understand evaluation metric (even if it is a curve), than a higher-order average which she/he has to pause a bit to ponder its meaning. What do you think?

Reply:
That's an important point, thanks for mentioning it!

I think you have something like [this plot in mind (p. 12, Fig. 9 & 10)](https://arxiv.org/pdf/2312.06254), where we have a time-based x-axis with equidistant retraining from a time trigger.

In Fig. 9, we see the different models produced by one pipeline run. 

For your research case of evaluating selection strategy that is totally fair, as you can use static triggering policies and adjust the evaluation strategies to this fixed schedules.

There are a few issues with this approach when comparing trigger policies though.

1) When comparing several trigger policies, we cannot rely on equidistant triggering timestamps, as different pipeline runs could have evaluations centered around different points in time. The `MatrixEvaluationStrategy` partially addresses this but does not focus on evaluating models only within the range where they are used for online inference.

2) We don't want to evaluate every model on all data (which is obvious) because it doesn’t matter how a model trained at the beginning of a dataset performs at the end since it will have been replaced by new models. Plotting this information adds noise without providing value when comparing different pipeline runs with different trigger policies. Cleaning up the data and combining it into one accuracy line is possible for a single dataset on one pipeline run, but comparing multiple pipeline runs is essential for Modyn. Thus, we need a method that allows for quick iterations on pipeline settings and is extensible.

3) Essentially, one can consider a pipeline run as a higher-order model that uses different trained models over time, resulting in an accuracy time series where one pipeline has one accuracy at a time. This time series for a fixed pipeline run is essentially an ensemble of the times series for the single models. We could consider the pipeline accuracy at a fixed time `t` as the accuracy of the model that is 'active' at time `t`.

4) Regarding your point on averaging, when comparing multiple pipeline runs with different trigger policies, we have at least three key dimensions to consider in plots: (dataset) time, cost (proxied by compute time), and evaluation metrics/performance (e.g., accuracy). To compare trigger policies based on their cost/accuracy trade-off, plotting 3 dimensionally is confusing therefore we need to aggregate over time. We could use a scatter plot with cost on the x-axis and accuracy on the y-axis, with one point per policy. This approach requires a scalar measure of accuracy for a pipeline.

Reply:
Overall, it would probably be possible to do everything with the current model evaluation strategies. I'm quite certain though that this would yield really intertwined code that is hard to adjust and extend.

Reply:
I see. Thanks for providing extra context! I will leave the discussion on how to eval in the triggering setting to you and Maxi

Reply:
I think all of these points are very valuable and should go into the document. In particular, before diving into the argument that we need to average, I think in this document we should explain that for the comparison of triggering policies, we need to reduce dimensionality and need to compress model performance into a scalar metric. In order to do this, we somehow need to reduce multiple accuracies (accuracy being any metric) that we collect into a single metric. I think we might want to start out by describing more background here in order to get a common understanding.

We get `n` model versions over time and evaluate those on `m` evaluation datasets (can be time slices of the dataset). So overall we have `n * m` values and need to reduce those. Basically it would be helpful to structure this section a bit differently, lay out what is happening and how we then need to process this data for your comparison. I think we explicitly want to talk about (what you also mention in your comments here) comparing selection policies with a common time trigger where we can plot the accuracies over time, and talk about comparing pipelines separately. Does this make sense?

Reply:
thanks for your thoughts @MaxiBoether! I will refine the document to give more background. Currently, there are, indeed, quite a lot of unstated assumptions.

Reply:
Thanks!